crates/sui-kvstore/src/bigtable/client.rs

```diff
@@ -0,0 +1,422 @@
+const EVENTS_COLUMN_QUALIFIER: &str = "ev";
+const TIMESTAMP_COLUMN_QUALIFIER: &str = "ts";
+const CHECKPOINT_NUMBER_COLUMN_QUALIFIER: &str = "cn";
+
+type Bytes = Vec<u8>;
+
+#[derive(Clone)]
+struct AuthChannel {
+    channel: Channel,
+    policy: String,
+    token_provider: Option<Arc<dyn TokenProvider>>,
+    token: Arc<RwLock<Option<Arc<Token>>>>,
+}
+
+#[derive(Clone)]
+pub struct BigTableClient {
+    table_prefix: String,
+    client: BigtableInternalClient<AuthChannel>,
+}
+
+#[async_trait]
+impl KeyValueStoreWriter for BigTableClient {
+    async fn save_objects(&mut self, objects: &[&Object]) -> Result<()> {
+        let mut items = Vec::with_capacity(objects.len());
+        for object in objects {
+            let object_key = ObjectKey(object.id(), object.version());
+            items.push((
+                Self::raw_object_key(&object_key)?,
+                vec![(DEFAULT_COLUMN_QUALIFIER, bcs::to_bytes(object)?)],
+            ));
+        }
+        self.multi_set(OBJECTS_TABLE, items).await
+    }
+
+    async fn save_transactions(&mut self, transactions: &[TransactionData]) -> Result<()> {
+        let mut items = Vec::with_capacity(transactions.len());
+        for transaction in transactions {
+            let cells = vec![
+                (
+                    TRANSACTION_COLUMN_QUALIFIER,
+                    bcs::to_bytes(&transaction.transaction)?,
+                ),
+                (
+                    EFFECTS_COLUMN_QUALIFIER,
+                    bcs::to_bytes(&transaction.effects)?,
+                ),
+                (EVENTS_COLUMN_QUALIFIER, bcs::to_bytes(&transaction.events)?),
+                (
+                    TIMESTAMP_COLUMN_QUALIFIER,
+                    bcs::to_bytes(&transaction.timestamp)?,
+                ),
+                (
+                    CHECKPOINT_NUMBER_COLUMN_QUALIFIER,
+                    bcs::to_bytes(&transaction.checkpoint_number)?,
+                ),
+            ];
+            items.push((transaction.transaction.digest().inner().to_vec(), cells));
+        }
+        self.multi_set(TRANSACTIONS_TABLE, items).await
+    }
+
+    async fn save_checkpoint(&mut self, checkpoint: &CheckpointData) -> Result<()> {
+        let summary = &checkpoint.checkpoint_summary.data();
+        let contents = &checkpoint.checkpoint_contents;
+        let signatures = &checkpoint.checkpoint_summary.auth_sig();
+        let key = summary.sequence_number.to_be_bytes().to_vec();
+        let cells = vec![
+            (CHECKPOINT_SUMMARY_COLUMN_QUALIFIER, bcs::to_bytes(summary)?),
+            (
+                CHECKPOINT_SIGNATURES_COLUMN_QUALIFIER,
+                bcs::to_bytes(signatures)?,
+            ),
+            (
+                CHECKPOINT_CONTENTS_COLUMN_QUALIFIER,
+                bcs::to_bytes(contents)?,
+            ),
+        ];
+        self.multi_set(CHECKPOINTS_TABLE, [(key, cells)]).await
+    }
+}
+
+#[async_trait]
+impl KeyValueStoreReader for BigTableClient {
```

Comment:
(For a later PR) Can we split up the read and write side of the KV store? It seems to me like these two sides are quite exclusive (readers don't write to KV stores and vice versa).

Reply:
It's already two different traits, and access is controlled by setting up a policy string. Readers won't be able to write to the Bigtable

Reply:
It was more from the perspective of building for me -- i.e. why build the logic to index when we're only ever going to read it and vice versa? We did the same thing in `sui-indexer` and it's not a big deal, but it is a source of friction.